components/brave_ads/core/internal/conversions/conversions_util.cc

```diff
@@ -1,96 +1,56 @@
-
-bool IsConversionIdValid(const std::string& conversion_id) {
-  return RE2::FullMatch(conversion_id, "^[a-zA-Z0-9-]*$");
-}
-
-}  // namespace
-
-std::string GetAlgorithm() {
-  return kAlgorithm;
-}
-
-absl::optional<VerifiableConversionEnvelopeInfo> SealEnvelope(
-    const VerifiableConversionInfo& verifiable_conversion) {
-  const std::string message = verifiable_conversion.id;
-  const std::string public_key_base64 = verifiable_conversion.public_key;
-
-  if (message.length() < kMinVerifiableConversionMessageLength ||
-      message.length() > kMaxVerifiableConversionMessageLength) {
-    return absl::nullopt;
+bool CanConvertAdEvent(const AdEventInfo& ad_event) {
+  // Only convert view-through and click-through ad events.
+  if (ad_event.confirmation_type != ConfirmationType::kViewed &&
+      ad_event.confirmation_type != ConfirmationType::kClicked) {
+    return false;
   }
 
-  if (!IsConversionIdValid(message)) {
-    return absl::nullopt;
-  }
+  switch (ad_event.type.value()) {
+    case AdType::kInlineContentAd:
+    case AdType::kPromotedContentAd: {
+      return UserHasOptedInToBraveNews();
+    }
 
-  // Protocol requires at least 2 trailing zero-padding bytes
-  std::vector<uint8_t> plaintext(message.cbegin(), message.cend());
-  plaintext.insert(plaintext.cend(), kCipherTextLength - plaintext.size(), 0);
-  CHECK_EQ(kCipherTextLength, plaintext.size());
+    case AdType::kNewTabPageAd: {
+      return UserHasOptedInToNewTabPageAds();
+    }
 
-  const absl::optional<std::vector<uint8_t>> public_key =
-      base::Base64Decode(public_key_base64);
-  if (!public_key) {
-    return absl::nullopt;
-  }
-  if (public_key->size() != crypto_box_PUBLICKEYBYTES) {
-    return absl::nullopt;
-  }
-
-  const crypto::KeyPairInfo ephemeral_key_pair = crypto::GenerateBoxKeyPair();
-  if (!ephemeral_key_pair.IsValid()) {
-    return absl::nullopt;
-  }
-
-  const std::vector<uint8_t> nonce = crypto::GenerateRandomNonce();
+    case AdType::kNotificationAd: {
+      return UserHasOptedInToBravePrivateAds();
+    }
 
-  const std::vector<uint8_t> padded_ciphertext = crypto::Encrypt(
-      plaintext, nonce, *public_key, ephemeral_key_pair.secret_key);
+    case AdType::kSearchResultAd: {
+      // Always.
+      return true;
+    }
 
-  // The first 16 bytes of the resulting ciphertext is left as padding by the
-  // C API and should be removed before sending out extraneously.
-  const std::vector<uint8_t> ciphertext(
-      padded_ciphertext.cbegin() + crypto_box_BOXZEROBYTES,
-      padded_ciphertext.cend());
-
-  VerifiableConversionEnvelopeInfo envelope;
-  envelope.algorithm = GetAlgorithm();
-  envelope.ciphertext = base::Base64Encode(ciphertext);
-  envelope.ephemeral_public_key =
-      base::Base64Encode(ephemeral_key_pair.public_key);
-  envelope.nonce = base::Base64Encode(nonce);
-
-  if (!envelope.IsValid()) {
-    return absl::nullopt;
+    case AdType::kUndefined: {
+      NOTREACHED_NORETURN();
+    }
```

Comment:
This looks like a false positive because `NOTREACHED_NORETURN()` will not return. Thanks

Reply:
@thypon @bcaller This is a false positive we should ideally filter out since `NOTREACHED_NORETURN` will automatically crash the browser if reached (i.e. no need for a `return` statement): https://source.chromium.org/chromium/chromium/src/+/main:base/notreached.h;l=49-50;drc=78ea882471b0219cbce7fe8020df8fa0245a3efe